cmd/checkpoint/main.go

```diff
@@ -47,49 +48,73 @@ func main() {
-	run(kubeAPIServer, tempAPIServer, api.NamespaceSystem, checkpoints)
+	run(api.NamespaceSystem, checkpoints)
 }
 
-func run(actualPodName, tempPodName, namespace string, checkpoints map[string]struct{}) {
+func run(namespace string, checkpoints map[string]struct{}) {
 	client := newAPIClient()
 	for {
-		_, checkpointed := checkpoints[checkpointManifest(tempPodName)]
-
 		var podList v1.PodList
 		if err := json.Unmarshal(getPodsFromKubeletAPI(), &podList); err != nil {
 			glog.Fatal(err)
 		}
-		switch {
-		case bothRunning(podList, actualPodName, tempPodName, namespace):
-			glog.Infof("both temp %v and actual %v pods running, removing temp pod", actualPodName, tempPodName)
-			// Both the temp and actual pods are running.
-			// Remove the temp manifest from the config dir so that the
-			// kubelet will stop it.
-			if err := os.Remove(activeManifest(tempPodName)); err != nil {
-				glog.Error(err)
+
+		pwc := getPodsWithCheckpointAnnotation(podList)
+		// special case for handling API server readiness
+		if _, ok := pwc[kubeAPIServer]; ok {
+			// Make sure it's actually running. Sometimes we get that
+			// pod manifest back, but the server is not actually running.
+			if _, err := client.Discovery().ServerVersion(); err != nil {
+				delete(pwc, kubeAPIServer)
 			}
-		case isPodRunning(podList, client, actualPodName, namespace):
-			glog.Infof("actual pod %v found, creating checkpoint pod manifest", actualPodName)
-			// The actual is running. Let's snapshot the pod,
-			// clean it up a bit, and then save it to the ignore path for
-			// later use.
-			checkpointPod := createCheckpointPod(podList, actualPodName, namespace)
-			convertSecretsToVolumeMounts(client, &checkpointPod)
-			writeManifest(checkpointPod, tempPodName)
-			checkpoints[checkpointManifest(tempPodName)] = struct{}{}
-			glog.Infof("finished creating checkpoint pod %v manifest at %s\n", tempPodName, checkpointManifest(tempPodName))
-
-		case checkpointed:
-			glog.Info("no actual pod running, installing checkpoint pod static manifest")
-			b, err := ioutil.ReadFile(checkpointManifest(tempPodName))
-			if err != nil {
-				glog.Error(err)
-			} else {
-				if err := ioutil.WriteFile(activeManifest(tempPodName), b, 0644); err != nil {
+		}
+
+		for name, pods := range pwc {
+			_, checkpointed := checkpoints[checkpointManifest(name)]
+
+			switch {
+			case len(pods) == 1 && !checkpointed:
+				glog.Infof("only actual pod %v found, creating checkpoint pod manifest", name)
+				// The actual is running. Let's snapshot the pod,
+				// clean it up a bit, and then save it to the ignore path for
+				// later use.
+				checkpointPod := createCheckpointPod(pods[0], name, namespace)
+				convertSecretsToVolumeMounts(client, &checkpointPod)
+				writeManifest(checkpointPod, name)
+				checkpoints[checkpointManifest(name)] = struct{}{}
+				glog.Infof("finished creating checkpoint pod %v manifest at %s\n", name, checkpointManifest(name))
+
+			case len(pods) == 2:
```

Comment:
Also RE: the api type meta -- that is just because we're talking to the kubelet api endpoint. Even static pods have that info if you get their status from the apiserver.